dircmp/dircmp.1
.\"
.\" SPDX-Licence-Identifier: CDDL-1.0
.\"
.\" Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany.
.\"
.\" Sccsid @(#)dircmp.1	1.1 (gritter) 6/25/05
.TH DIRCMP 1 "6/25/05" "Heirloom Toolchest" "User Commands"
.SH NAME
dircmp \- directory comparison
.SH SYNOPSIS
\fBdircmp\fR [\fB\-ds\fR] [\fB\-w\fI width\fR] \fIdirectory directory\fR
.SH DESCRIPTION
The
.I dircmp
command creates a list of files which are
identical, different, or exist in only one of the two
.I directories
given.
.PP
With the
.B \-d
option, differing files are compared using
.IR diff (1),
and its output is included.
.PP
The
.B \-s
option suppresses messages about identical files.
.PP
Output is paginated through
.IR pr (1).
The
.B \-w
option specifies a maximum
.I width
for the formatted output.
.SH "SEE ALSO"
diff(1)
.SH NOTES
.I dircmp
ignores files whose names contain backslash or newline characters.


CHANGES
Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.

dircmp/dircmp.sh
#
# SPDX-Licence-Identifier: CDDL-1.0
#
#	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T
#	  All Rights Reserved

#
#	Copyright (c) 1996-2000 by Sun Microsystems, Inc.
#	All rights reserved.
# from OpenSolaris "dircmp.sh	1.21	05/06/08 SMI"
#     Portions Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany
#     Sccsid @(#)dircmp.sh	1.6 (gritter) 7/1/05
tmpdir=/var/tmp
PATH=@SV3BIN@:@DEFBIN@:$PATH export PATH
progname=`basename $0`
USAGE="$progname: usage: $progname -s -d -wn directory directory"
trap "rm -f $tmpdir/dc$$*;exit" 1 2 3 15
exitstat=0
sizediff=
cmpdiff=
Sflag=0
Dflag=0
fsize1=
fsize2=
LFBOUND=2147483648
width=72

#
# function to generate consistent "diff" output whether or not files are intact
#
dodiffs() {
	type=`LC_ALL=C file "$D1/$a"`
	case "$type" in
		*text)  ;;
		*script) ;;
		*empty*) echo $D1/`basename "$a"` is an empty file |
			  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
        	*cannot*) echo $D1/`basename "$a"` does not exist |
			  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
        	*)	echo $D1/`basename "$a"` is an object file |
		   	  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
	esac
	type=`LC_ALL=C file "$D2/$a"`
	case "$type" in
        	*text)  ;;
        	*script) ;;
        	*empty*) echo $D2/`basename "$a"` is an empty file |
			  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
        	*cannot*) echo $D2/`basename "$a"` does not exist |
			  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
        	*)	echo $D2/`basename "$a"` is an object file |
			  pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
			continue
        	;;
	esac
	#
	# If either is a "large file" use bdiff (LF aware),
	# else use diff.
	#
	if test $fsize1 -lt $LFBOUND -a $fsize2 -lt $LFBOUND
	then cmd="diff"
	else cmd="bdiff"
	fi
	($cmd "$D1/$a" "$D2/$a"; echo $? > $tmpdir/dc$$status) | \
	    pr -h "diff of $a in $D1 and $D2" >> $tmpdir/dc$$g
	if test `cat $tmpdir/dc$$status` -ne 0
	then exitstat=$diffstat
	fi
}
#
# dircmp entry point
#
while getopts dsw: i
do
	case $i in
	d)	Dflag=1;;
	s)	Sflag=1;;
	w)	width=`expr $OPTARG + 0 2>/dev/null`
		if [ $? = 2 ]
		then echo "$progname: numeric argument required"
			exit 2
		fi
		;;
	*)	echo $USAGE
		exit 2;;
	esac
done
shift `expr $OPTIND - 1`
#
D0=`pwd`
D1=$1
D2=$2
if [ $# -lt 2 ]
then echo $USAGE
     exit 1
elif [ ! -d "$D1" ]
then echo $D1 not a directory !
     exit 2
elif [ ! -d "$D2" ]
then echo $D2 not a directory !
     exit 2
fi
#
# find all dirs/files in both directory hierarchies. Use "comm" to identify
# which are common to both hierarchies as well as unique to each.
# At this point, print those that are unique.
#
cd "$D1"
find . -name '*[
\\]*' -prune -o ! -name '*[
\\]*' -print | sort > $tmpdir/dc$$a
cd "$D0"
cd "$D2"
find . -name '*[
\\]*' -prune -o ! -name '*[
\\]*' -print | sort > $tmpdir/dc$$b
comm $tmpdir/dc$$a $tmpdir/dc$$b | sed -n \
	-e "/^		/w $tmpdir/dc$$c" \
	-e "/^	[^	]/w $tmpdir/dc$$d" \
	-e "/^[^	]/w $tmpdir/dc$$e"
rm -f $tmpdir/dc$$a $tmpdir/dc$$b
pr -w${width} -h "$D1 only and $D2 only" -m $tmpdir/dc$$e $tmpdir/dc$$d
rm -f $tmpdir/dc$$e $tmpdir/dc$$d
#
# Generate long ls listings for those dirs/files common to both hierarchies.
# Use -lgn to avoid problem when user or group names are too long, causing
# expected field separator to be missing
# Avoid other potential problems by piping through sed:
#  - Remove: Spaces in size field for block & character special files
#      '71, 0' becomes '710'
#  - For file name, do not print '-> some_file'
#      '/tmp/foo -> FOO' becomes '/tmp/foo'

# The following sed is to read filenames with special characters
sed -e 's/..//'  -e  's/\([^-a-zA-Z0-9/_.]\)/\\\1/g' < $tmpdir/dc$$c > $tmpdir/dc$$f


cat $tmpdir/dc$$f | xargs ls -lLgnd | \
  sed -e '/^[bc]/ s/, *//' -e '/^l/ s/ -> .*//' > $tmpdir/dc$$i 2>/dev/null
cd "$D0"
cd "$D1"


cat $tmpdir/dc$$f | xargs ls -lLgnd | \
sed -e '/^[bc]/ s/, *//' -e '/^l/ s/ -> .*//' > $tmpdir/dc$$h 2>/dev/null
cd "$D0"
> $tmpdir/dc$$g
#
# Process the results of the 'ls -lLgnd' to obtain file size info
# and identify a large file's existence.
#
while read <&3 tmp tmp tmp fsize1 tmp tmp tmp a &&
      read <&4 tmp tmp tmp fsize2 tmp tmp tmp b; do
	#
	# A window of opportunity exists where the ls -lLgnd above
	# could produce different
	# results if any of the files were removed since the find command.
	# If the pair of reads above results in different values (file names) for 'a'
	# and 'b', then get the file pointers in sync before continuing, and display
	# "different" message as customary.
	#
	if test "x$a" != "x$b"; then
	while test "$a" -lt "$b"; do
		if test $Sflag -ne 1
		then echo "different	$a"
		dodiffs
		fi
		read <&3 tmp tmp tmp fsize1 tmp tmp tmp a
	done
	while test "$a" -gt "$b"; do
		if test $Sflag -ne 1
		then echo "different	$b"
		dodiffs
		fi
		read <&4 tmp tmp tmp fsize2 tmp tmp tmp b
	done
	fi
	cmpdiff=0
	sizediff=0
	if [ -d "$D1/$a" ]
	then if test $Sflag -ne 1
	     then echo "directory	$a"
	     fi
	elif [ -f "$D1/$a" ]
	then
	     #
	     # If the file sizes are different, then we can skip the run
	     # of "cmp" which is only used to determine 'same' or 'different'.
	     # If the file sizes are the same, we still need to run "cmp"
	     #
	     if test $fsize1 -ne $fsize2
	     then
		sizediff=1
	     else
		cmp -s -- "$D1/$a" "$D2/$a"
		cmpdiff=$?
	     fi
	     if test $sizediff -eq 0 -a $cmpdiff -eq 0
	     then if test $Sflag -ne 1
		  then echo "same     	$a"
		  fi
	     else echo "different	$a"
		  if test $Dflag -eq 1
		  then
			dodiffs
		  fi
	     fi
	elif test $Sflag -ne 1
	then echo "special  	$a"
	fi
done 3<$tmpdir/dc$$h 4<$tmpdir/dc$$i | pr -r -h "Comparison of $D1 $D2"
if test $Dflag -eq 1
then cat $tmpdir/dc$$g
fi
rm -f $tmpdir/dc$$*
exit $exitstat
